### PR TITLE
US020 Delete Role

### PR DESCRIPTION
**Pull request by eoinmcbennett**

### CODE REVIEW

**Review by dianemorrison — APPROVED** [review]

File: src/main/java/org/kainos/ea/api/JobService.java [hunk 1]
```diff
@@ -28,5 +30,30 @@ public List<Job> getAllJobs() throws FailedToGetJobsException {
...
+    public Job getById(int id) throws FailedToGetJobsException {
+        try {
+            return jobDao.getJobById(databaseConnector.getConnection(),id);
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToGetJobsException();
+        }
+    }
+
+    public boolean delete(int id) throws FailedtoDeleteException {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/JobService.java]: don't need to return boolean, just use void

File: src/main/java/org/kainos/ea/api/JobService.java [hunk 2]
```diff
@@ -28,5 +30,30 @@ public List<Job> getAllJobs() throws FailedToGetJobsException {
...
+        try {
+
+            Job job = jobDao.getJobById(databaseConnector.getConnection(),id);
+            if (job == null) {
+                System.err.println("Attempted to delete job by id " + id + " but it doesn't exist!");
+                return false;
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/JobService.java]: throw exception if this doesn't exist

File: src/main/java/org/kainos/ea/db/JobDao.java [hunk 3]
```diff
@@ -43,9 +44,72 @@ public List<Job> getAllJobs(Connection connection) throws SQLException {
...
+        if (connection == null) {
+            System.err.println("Failed to connect to database");
+            throw new SQLException("Failed to connect to database");
+        }
 
+
+        if(connection.isClosed()){
+            System.err.println("Database connection was closed");
+            throw new SQLException("Database connection was closed");
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/db/JobDao.java]: this logic should be handles by the database connector

File: src/main/java/org/kainos/ea/db/JobDao.java [hunk 4]
```diff
@@ -43,9 +44,72 @@ public List<Job> getAllJobs(Connection connection) throws SQLException {
...
+        if (connection == null) {
+            System.err.println("Failed to connect to database");
+            throw new SQLException("Failed to connect to database");
+        }
+
+
+        if(connection.isClosed()){
+            System.err.println("Database connection was closed");
+            throw new SQLException("Database connection was closed");
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/db/JobDao.java]: this logic should be handles by the database connector

File: src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java [hunk 5]
```diff
@@ -49,4 +51,40 @@ void getAllJobs_shouldReturn500StatusCodeAndErrorMessage_whenServiceThrowsFailed
         Assertions.assertEquals(response.getEntity(), "A database error occurred, failed to get jobs");
          }
 
+    @Test
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java:65]: no happy path unit test here

File: src/test/java/org/kainos/ea/service/JobServiceUnitTest.java [hunk 6]
```diff
@@ -59,9 +60,37 @@ void getAllJobs_shouldThrowFailedToGetJobsException_whenDaoThrowsSQLException()
...
 
         assertThrows(FailedToGetJobsException.class,
                 () -> jobService.getAllJobs());
     }
 
+    @Test
+    void deleteJob_shouldReturnFalse_IfJobIdNotFound() throws SQLException, FailedtoDeleteException {
+        Mockito.when(databaseConnector.getConnection()).thenReturn(connection);
+        Mockito.when(jobDao.getJobById(connection,-1)).thenReturn(null);
+        Mockito.doNothing().when(jobDao).deleteJob(isA(Connection.class),isA(Integer.class));
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobServiceUnitTest.java:75]: you don't need to do this, it'll do nothing by default

File: src/test/java/org/kainos/ea/service/JobServiceUnitTest.java [hunk 7]
```diff
@@ -59,9 +60,37 @@ void getAllJobs_shouldThrowFailedToGetJobsException_whenDaoThrowsSQLException()
...
+        assertFalse(jobService.delete(-1));
+    }
+
+    @Test
+    void deleteJob_shouldReturnTrue_IfJobIdFound() throws SQLException, FailedtoDeleteException {
+        Job sampleJob1 = new Job(1, "Software Engineer","Capbility", "Develops, tests, and maintains software applications, collaborates with cross-functional teams, and follows best practices to deliver efficient and reliable software solutions.", "https://kainossoftwareltd.sharepoint.com/people/Job%20Specifications/Forms/AllItems.aspx?id=%2Fpeople%2FJob%20Specifications%2FEngineering%2FJob%20profile%20%2D%20Software%20Engineer%20%28Associate%29%2Epdf&parent=%2Fpeople%2FJob%20Specifications%2FEngineering&p=true&ga=1","Trainee");
+        Mockito.when(databaseConnector.getConnection()).thenReturn(connection);
+        Mockito.when(jobDao.getJobById(connection,-1)).thenReturn(sampleJob1);
+        Mockito.doNothing().when(jobDao).deleteJob(isA(Connection.class),isA(Integer.class));
+        assertFalse(jobService.delete(1));
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobServiceUnitTest.java]: this asserts false which isn't what the method name suggests

**Review by shaunganley — CHANGES_REQUESTED** [review]
needs integration tests

File: src/main/java/org/kainos/ea/api/JobService.java (hunk 1 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/api/JobService.java]: Converted to use exceptions

File: src/main/java/org/kainos/ea/db/JobDao.java (hunk 3 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/JobDao.java]: Removed

File: src/main/java/org/kainos/ea/db/JobDao.java (hunk 4 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/db/JobDao.java]: Removed

File: src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java (hunk 5 above)
Comment (eoinmcbennett) [reply src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java:65]: Sorted

File: src/test/java/org/kainos/ea/service/JobServiceUnitTest.java (hunk 6 above)
Comment (eoinmcbennett) [reply src/test/java/org/kainos/ea/service/JobServiceUnitTest.java:75]: Method was changed to throw not found exception, test changed to reflect this

File: src/test/java/org/kainos/ea/service/JobServiceUnitTest.java (hunk 7 above)
Comment (eoinmcbennett) [reply src/test/java/org/kainos/ea/service/JobServiceUnitTest.java]: Changed to check for nothing being thrown to align with new method behaviour.

File: src/main/java/org/kainos/ea/api/JobService.java (hunk 2 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/api/JobService.java]: Changed to do this

File: src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java [hunk 8]
```diff
@@ -65,4 +71,25 @@ void getAllJobs_ShouldReturnListOfJobs() {
 
     }
 
+    @Test
+    void deleteJob_shouldDeleteJob() throws SQLException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java:75]: add an integration test to delete a job that doesn't exist and expect 404

**Review by shaunganley — CHANGES_REQUESTED** [review]
needs integration and unit tests for get job endpoint

**Review by shaunganley — APPROVED** [review]

**Pull request merged**